### PR TITLE
i#4031: fix assert on zero-length memory range passed to `add_vm_area`

### PR DESCRIPTION
**Pull request by petrochenkov**
Fixes: #4031

### CODE REVIEW

**Comment by petrochenkov** [comment]
I'm not sure how to test anything that fails only with Debug build of DynamoRIO, like this assert.
The added test passed in Release mode even before the fix.

File: suite/tests/linux/zero-length-mem-ranges.c [hunk 1]
```diff
@@ -0,0 +1,40 @@
+#define _GNU_SOURCE
```
Comment (derekbruening) [inline suite/tests/linux/zero-length-mem-ranges.c]: Please include the copyright + license header

File: suite/tests/linux/zero-length-mem-ranges.c [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        return EXIT_FAILURE;
+    }
+
+    int munmap_res = munmap(mem, 0);
+    if (munmap_res != -1) {
+        printf("zero-length munmap succeeded\n");
+        return EXIT_FAILURE;
+    }
+
+    printf("done\n");
```
Comment (derekbruening) [inline suite/tests/linux/zero-length-mem-ranges.c]: We always #include "tools.h" and then use `print()` for standardized flushing behavior across platforms and interleaving w/ DR output (no other output here now but the test could be run under a client in the future or sthg).

So s/printf/print/g.

**Comment by derekbruening** [comment]
Please also fix the CI failures (click on the red X) -- looks like DR reports a memory leak when running this test on multiple platforms so probably you can repro it?

**Comment by derekbruening** [comment]
run arm tests

File: suite/tests/linux/zero-length-mem-ranges.c (hunk 1 above)
Comment (petrochenkov) [reply suite/tests/linux/zero-length-mem-ranges.c]: Done.

File: suite/tests/linux/zero-length-mem-ranges.c (hunk 2 above)
Comment (petrochenkov) [reply suite/tests/linux/zero-length-mem-ranges.c]: Done.

**Comment by petrochenkov** [comment]
I moved the zero-length check to a different place (munmap post-syscall) to avoid the memory leak, looks like the assert isn't hit from any other places.

**Comment by derekbruening** [comment]
> I moved the zero-length check to a different place (munmap post-syscall) to avoid the memory leak, looks like the assert isn't hit from any other places.

The leak is in DR's memory so I assume that DR is allocating something, then short-circuiting or exiting due to zero length, and never freeing that something: if you look at the code handling the app code that you changed to avoid the leak and you can spot the bug maybe you could fix it in DR if it's simple?

**Comment by petrochenkov** [comment]
>if you look at the code handling the app code that you changed to avoid the leak and you can spot the bug maybe you could fix it in DR if it's simple?

There's too much happening between `post_system_call(SYS_munmap)` and `add_vm_area`, I looked through the code, but still don't understand most of it, sorry.

There's one suspicious thing though, the first thing we do in post-munmap is `query_memory_ex_from_os(addr, &info);`, which is pretty strange for zero-length ranges because the `addr` is not even in the range in this case.
Perhaps all the further nonsense starts with this arbitrary filled `dr_mem_info_t`? In that case this PR's solution would actually be correct.

**Comment by petrochenkov** [comment]
The AppVeyor failure looks unrelated to this PR.

**Comment by derekbruening** [comment]
> > if you look at the code handling the app code that you changed to avoid the leak and you can spot the bug maybe you could fix it in DR if it's simple?
> 
> There's too much happening between `post_system_call(SYS_munmap)` and `add_vm_area`, I looked through the code, but still don't understand most of it, sorry.
> 
> There's one suspicious thing though, the first thing we do in post-munmap is `query_memory_ex_from_os(addr, &info);`, which is pretty strange for zero-length ranges because the `addr` is not even in the range in this case.
> Perhaps all the further nonsense starts with this arbitrary filled `dr_mem_info_t`? In that case this PR's solution would actually be correct.

So I assume the DR leak is unrelated to this PR (i.e., if you remove the add_vm_area fix here and run the original test (I guess disabling the assert to get to the leak check at the end in debug), the leak is still reported, right?  Assuming so, if you could file a separate issue on it and include the part of the test that triggers it, and then we'll merge this PR for add_vm_area.

**Comment by petrochenkov** [comment]
>i.e., if you remove the add_vm_area fix here and run the original test (I guess disabling the assert to get to the leak check at the end in debug), the leak is still reported, right?

No the leak is not reported in that case.
The leak is indeed related to this PR in its original form (https://github.com/DynamoRIO/dynamorio/pull/4235/commits/ff9f707f38deb54de4504f9a26c6a9566c8f191f).
If the body of `add_vm_area` is skipped, but the rest of code in `post_system_call(SYS_munmap)` is not, then we get the leak, but I don't understand why.
So the current version of this PR skips all the `post_system_call(SYS_munmap)`.

File: core/unix/os.c [hunk 3]
```diff
@@ -8178,6 +8178,8 @@ post_system_call(dcontext_t *dcontext)
         /* We assumed in pre_system_call() that the unmap would succeed
          * and flushed fragments and removed the region from exec areas.
          * If the unmap failed, we re-add the region to exec areas.
+         * For zero-lenght unmaps we don't need to re-add anything,
```
Comment (derekbruening) [inline core/unix/os.c:8181]: spelling: length

**Comment by derekbruening** [comment]
run arm tests

**Comment by derekbruening** [comment]
run arm tests

File: core/unix/os.c (hunk 3 above)
Comment (derekbruening) [reply core/unix/os.c]: It let me fix it so I went ahead and did that and will merge after the tests run.
Thanks again for the patch!

**Comment by derekbruening** [comment]
Hmm, Appveyor is the drwap-test-detach in 32-bit: seems to timeout.  Under #4058.

**Pull request merged**